### PR TITLE
Fixes #28204 - Run Dynflow with Sidekiq using docker-compose

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

**Comment by adamruzicka** [comment]
@ohadlevy delivering what I promised :)

File: docker-compose.yml [hunk 1]
```diff
@@ -32,6 +32,7 @@ services:
       - FOREMAN_DOMAIN=example.com
       - FOREMAN_RAILS_CACHE_STORE_TYPE=redis
       - FOREMAN_RAILS_CACHE_STORE_URLS=redis://redis:6379/0
+      - REDIS_URL=redis://redis:6379/1
```
Comment (ohadlevy) [inline docker-compose.yml]: should we expose it as a DYNFLOW or SIDEKIQ prefix?

Comment (adamruzicka) [reply docker-compose.yml]: We would then have to modify our code. As it is now `REDIS_URL` gets read/set/used by Sidekiq itself, not by us

EDIT: I'm not opposed to this, just saying.

File: docker-compose.yml [hunk 2]
```diff
@@ -45,16 +46,16 @@ services:
...
     <<: *app_base
-    command: bundle exec rake dynflow:executor
-    hostname: worker.example.com
-    ports:
-      - 9999
-    healthcheck:
-      test: "ps ax | grep -v grep | grep dynflow_executor"
-      interval: 1m
-      start_period: 1m
+    command: bundle exec sidekiq -r ./extras/dynflow-sidekiq.rb -c 15 -q default,rex
```
Comment (ezr-ondrej) [inline docker-compose.yml]: Isn't the rex queue full named `remote_execution` instead of `rex`?

Comment (adamruzicka) [reply docker-compose.yml]: It is, thank you for spotting that

**Comment by ohadlevy** [comment]
I'll test this as soon as I can, I just learned that docker-compose doesn't work easily with fedora 31...

**Comment by ohadlevy** [comment]
based on the conversation at https://community.theforeman.org/t/rfc-redis-in-foreman/16102/3 i believe this needs to be updated to reflect that?

**Comment by ezr-ondrej** [comment]
[test katello]

File: docker-compose.yml [hunk 3]
```diff
@@ -31,11 +31,13 @@ services:
       - FOREMAN_FQDN=foreman.example.com
       - FOREMAN_DOMAIN=example.com
       - FOREMAN_RAILS_CACHE_STORE_TYPE=redis
-      - FOREMAN_RAILS_CACHE_STORE_URLS=redis://redis:6379/0
+      - FOREMAN_RAILS_CACHE_STORE_URLS=redis://redis-cache:6379/0
+      - DYNFLOW_REDIS_URL=redis://redis-tasks:6379/0
```
Comment (ezr-ondrej) [inline docker-compose.yml:35]: It seems `redis-tasks` got translated to `127.0.0.1` for some reason and it's not connecting to the image.

**Review by ezr-ondrej — COMMENTED** [review]
The dynflow processes are unable to connect to the redis.
For some reason I have an executor as part of the main app.

The sidekiq executor is visible in the dynflow console though. I just don't know if it's processing anything.

**Comment by adamruzicka** [comment]
Interesting

```
$ bundle exec sidekiq -r ./extras/dynflow-sidekiq.rb -c 1 -q dynflow_orchestrator                                                                                                                                                   
2019-12-10T13:19:25.275Z 10 TID-gr1sem6pq INFO: GitLab reliable fetch activated!                                                                                                                                                            
2019-12-10T13:19:25.275Z 10 TID-gr1sneioi INFO: Booting Sidekiq 5.2.7 with redis options {:id=>"Sidekiq-server-PID-10", :url=>nil}                                                                                                          
2019-12-10T13:19:25.276Z 10 TID-gr1sneioi ERROR: Heartbeat thread error: Error connecting to Redis on 127.0.0.1:6379 (Errno::ECONNREFUSED)                                                                                                  
2019-12-10T13:19:26.297Z 10 TID-gr1sneioi ERROR: Heartbeat thread error: Error connecting to Redis on 127.0.0.1:6379 (Errno::ECONNREFUSED)                                                                                                  
2019-12-10T13:19:27 [I|app|] Rails cache backend: Redis                                                               
2019-12-10T13:19:27.300Z 10 TID-gr1sneioi ERROR: Heartbeat thread error: Error connecting to Redis on 127.0.0.1:6379 (Errno::ECONNREFUSED)
2019-12-10T13:19:28.333Z 10 TID-gr1sneioi ERROR: Heartbeat thread error: Error connecting to Redis on 127.0.0.1:6379 (Errno::ECONNREFUSED)                                                                                                  
2019-12-10T13:19:29 [W|app|] Creating scope :completer_scope. Overwriting existing method Location.completer_scope.   
2019-12-10T13:19:29 [W|app|] Creating scope :completer_scope. Overwriting existing method Organization.completer_scope.                                                                                                                     
2019-12-10T13:19:29.337Z 10 TID-gr1sneioi ERROR: Heartbeat thread error: Error connecting to Redis on 127.0.0.1:6379 (Errno::ECONNREFUSED)                                                                                                  
2019-12-10T13:19:30.344Z 10 TID-gr1sneioi ERROR: Heartbeat thread error: Error connecting to Redis on 127.0.0.1:6379 (Errno::ECONNREFUSED)
2019-12-10T13:19:31.345Z 10 TID-gr1sneioi ERROR: Heartbeat thread error: Error connecting to Redis on 127.0.0.1:6379 (Errno::ECONNREFUSED)                                                                                                  
2019-12-10T13:19:32.947Z 10 TID-gr1sneioi ERROR: Heartbeat thread error: Error connecting to Redis on 127.0.0.1:6379 (Errno::ECONNREFUSED)                                                                                                  
redis://redis-tasks:6379/0                                                                                                                                                                                                                  
2019-12-10T13:19:33.496Z 10 TID-gr1sem6pq INFO: Booting Sidekiq 5.2.7 with redis options {:url=>"redis://redis-tasks:6379/0", :id=>"Sidekiq-server-PID-10"}                                                                                 
2019-12-10T13:19:33 [I|dyn|] Acquired orchestrator lock, entering active mode.                                                                                                                                                              
2019-12-10T13:19:35 [I|app|] Everything ready for world: 4794d16e-0bb4-497c-b8ea-5fe1a994063e                                                                                                                                               
2019-12-10T13:19:35.381Z 10 TID-gr1sem6pq INFO: Running in ruby 2.6.3p62 (2019-04-16 revision 67580) [x86_64-linux]                                                                                                                         
2019-12-10T13:19:35.382Z 10 TID-gr1sem6pq INFO: See LICENSE and the LGPL-3.0 for licensing details.
2019-12-10T13:19:35.382Z 10 TID-gr1sem6pq INFO: Upgrade to Sidekiq Pro for more features and support: http://sidekiq.org
2019-12-10T13:19:35.382Z 10 TID-gr1sem6pq INFO: Starting processing, hit Ctrl-C to stop
2019-12-10T13:19:35.384Z 10 TID-gr1u762pq Dynflow::Executors::Sidekiq::OrchestratorJobs::WorkerDone JID-6d2fe7f67a1bb2f5dff94963 INFO: start
2019-12-10T13:19:35.385Z 10 TID-gr1u762pq Dynflow::Executors::Sidekiq::OrchestratorJobs::WorkerDone JID-6d2fe7f67a1bb2f5dff94963 INFO: done: 0.001 sec
```

File: extras/dynflow-sidekiq.rb [hunk 4]
```diff
@@ -8,9 +8,6 @@
 rails_env_file = File.expand_path('./config/environment.rb', rails_root)
 require rails_env_file
 
-redis_url = ENV['DYNFLOW_REDIS_URL'] || SETTINGS.dig(:dynflow, :redis_url)
```
Comment (ezr-ondrej) [inline extras/dynflow-sidekiq.rb]: I am not sure I understand (dig xD) why you have done this? how we should configure redis url now? :(

File: docker-compose.yml (hunk 3 above)
Comment (ohadlevy) [reply docker-compose.yml:35]: how did you execute it?

File: extras/dynflow-sidekiq.rb [hunk 5]
```diff
@@ -8,9 +8,6 @@
 rails_env_file = File.expand_path('./config/environment.rb', rails_root)
 require rails_env_file
 
-redis_url = SETTINGS.dig(:dynflow, :redis_url)
```
Comment (ohadlevy) [inline extras/dynflow-sidekiq.rb:11]: is this neded? all settings afair first read from env variables.. am i mistaken @timogoebel ?

File: extras/dynflow-sidekiq.rb (hunk 4 above)
Comment (adamruzicka) [reply extras/dynflow-sidekiq.rb]: Only using env variables (the last commit also changed the systemd unit). Alternatively we would need to set up reliable fetch in dynflow after the url was properly configured.

File: extras/dynflow-sidekiq.rb (hunk 5 above)
Comment (timogoebel) [reply extras/dynflow-sidekiq.rb:11]: The settings have to be registered, see #6826.

File: config/settings.yaml.example [hunk 6]
```diff
@@ -131,7 +131,6 @@
 #
 # :dynflow:
 #   :pool_size: 5
```
Comment (ekohl) [inline config/settings.yaml.example:133]: If you're removing the redis URL, do we want to move the pool size as well?

File: docker-compose.yml (hunk 3 above)
Comment (ezr-ondrej) [reply docker-compose.yml:35]: `docker-compose build` `docker-compose run migration, seed` `docker-compose up` :thinking:

Comment (adamruzicka) [reply docker-compose.yml:35]: It didn't get translated per-se. Sidekiq reliable fetch tried to connect to redis before we could configure sidekiq to use the value from this env var.

**Comment by ohadlevy** [comment]
tested locally and works fine (using podman-compose), rebased and pushed to re-run tests.

**Comment by ohadlevy** [comment]
test failures not related - thanks!

**Pull request merged**